### ISSUE TITLE
Excessive memory usage on Kubernetes

### ISSUE THREAD

**Issue by akelge** [fluent/fluent-bit#7726]
## Bug Report

**Describe the bug**
Memory usage of fluent-bit on Kubernetes goes sky high until either uses the whole memory or gets OOM killed.
We are talking of 1/2Gb memory on a quite quiet cluster

**To Reproduce**

Just let fluent-bit run

**Expected behavior**

fluent-bit should use less memory, with the simple configuration we are using.

**Screenshots**
<img width="1838" alt="image" src="https://github.com/fluent/fluent-bit/assets/7236218/38ad80fa-81cd-4a27-83f4-1f6a87376f80">
<img width="531" alt="image" src="https://github.com/fluent/fluent-bit/assets/7236218/8b5a992c-f1c3-4263-b541-3c03b310d7e9">

**Your Environment**

* Version used: 2.1.6
* Configuration: attached
* Environment name and version: kubernetes 1.24, deployed via offical helm-chart
* Server type and version: RKE2
* Operating System and version: n/a
* Filters and plugins: see attached configuration

**Additional context**
After some investigation it seems that fluent-bit, per se, is using the correct amount of memory, but that the container is using a lot of buffers. When memory approaches the limit, if I am lucky, buffers are released and fluent-bit remains ok, else it is OOM killed.

Based on https://github.com/aws-samples/amazon-ecs-firelens-examples/tree/mainline/examples/fluent-bit/oomkill-prevention we calculated an expected memory usage of 128Mb and set the memory limit to 320Mb to play it safe. Actual fluent-bit memory usage seems to be around 128Mb, but the whole container goes up to 320+

Here is the configuration I am using:

```
[SERVICE]
    Daemon Off
    Flush 1
    Log_Level info
    HTTP_Server On
    HTTP_Listen 0.0.0.0
    HTTP_Port 2020
    Parsers_File parsers.conf
    Parsers_File custom_parsers.conf
    Health_Check On
    storage.path /storage
    storage.metrics On
    storage.backlog.mem_limit 5M
    storage.checksum Off
    storage.sync normal
    storage.delete_irrecoverable_chunks On
    Mem_Buf_Limit 5M
    storage.max_chunks_up 16

[INPUT]
    Name tail
    Path /var/log/containers/*.log
    Exclude_Path /var/log/containers/*kube-system*,/var/log/containers/*monitoring*,/var/log/containers/*longhorn-system*
    DB /storage/containers.db
    Parser cri
    Multiline.parser cri
    Tag kube.*
    Mem_Buf_Limit 5MB
    Skip_Long_Lines On
    Skip_Empty_Lines On
    Ignore_Older 48h
    storage.type  filesystem
    storage.pause_on_chunks_overlimit on

[INPUT]
    Name tail
    Path /var/log/containers/*ingress-nginx-controller*.log
    DB /storage/system.db
    Parser cri
    Multiline.parser cri
    Tag kube.*
    Mem_Buf_Limit 5MB
    Skip_Long_Lines On
    Skip_Empty_Lines On
    Ignore_Older 48h
    storage.type  filesystem
    storage.pause_on_chunks_overlimit on


[FILTER]
    Name              grep
    Match             kube.*
    Exclude           log ^\s*$


[FILTER]
    Name kubernetes
    Match kube.*

    Merge_Log Off
    Keep_Log Off
    Use_Kubelet On
    Buffer_Size 0
    K8S-Logging.Parser On
    K8S-Logging.Exclude On

[FILTER]
    Name                   nest
    Match                  kube.*

    Operation              lift
    Nested_under           kubernetes
    Add_prefix             kubernetes_

[FILTER]
    Name                  nest
    Match                 kube.*

    Operation             lift
    Nested_under          kubernetes_labels
    Add_prefix            kubernetes_labels_

[FILTER]
    Name                  nest
    Match                 kube.*

    Operation             lift
    Nested_under          kubernetes_annotations
    Add_prefix            kubernetes_annotations_

[FILTER]
    Name    modify
    Match   kube.*
    Condition Key_Exists level
    Condition Key_Value_Equals level warn
    Remove level
    Set level warning

[FILTER]
    Name    modify
    Match   kube.*

    Copy kubernetes_namespace_name namespace
    Copy kubernetes_container_name container
    Copy kubernetes_container_image container_image
    Copy kubernetes_pod_name pod

    Copy kubernetes_labels_deployment deployment
    Copy kubernetes_labels_app app
    Copy kubernetes_labels_environment environment
    Copy kubernetes_labels_logging-selector logging_selector
    Copy kubernetes_labels_logging-selector logging-selector

[FILTER]
    Name nest
    Match kube.*

    Operation nest
    Wildcard kubernetes_*
    Nest_Under kubedrop

[FILTER]
    Name    record_modifier
    Match   kube.*

    Record  cluster CLUSTERNAME
    Remove_key  _p
    Remove_key  kubedrop

[OUTPUT]
    Name gelf
    Match *
    Host graylog-gelf-udp.graylog
    Port 12201
    Gelf_Host_Key cluster
    Gelf_Short_Message_Key log
    Mode udp
    Compress true
    Retry_Limit false
    storage.total_limit_size  5M
```

**Comment by nokute78** [comment]
out_gelf caused memory leak.
It was fixed by https://github.com/fluent/fluent-bit/pull/7640 and merged from v2.1.7.

Could you try v2.1.7+ ?

**Comment by akelge** [comment]
I installed the latest version and indeed now it is working as expected. Memory grows, but then it is reclaimed and overall memory usage is what I was expecting.

Resolved!
